ambry-protocol/src/main/java/com/github/ambry/protocol/GetResponse.java

```diff
@@ -114,46 +122,111 @@ public static GetResponse readFrom(DataInputStream stream, ClusterMap map) throw
    * {@link GetResponse#writeTo(AsyncWritableChannel, Callback)}.
    * This method allocate bufferToSend and write metadata to it if bufferToSend is null.
    */
-  private void prepareBufferToSend() {
-    if (bufferToSend == null) {
-      bufferToSend = ByteBuffer.allocate(
+  @Override
+  protected void prepareBuffer() {
+    if (toSend != null && toSend.content() != null) {
+      bufferToSend = PooledByteBufAllocator.DEFAULT.ioBuffer((int) sizeInBytes());
+    } else {
+      bufferToSend = PooledByteBufAllocator.DEFAULT.ioBuffer(
           (int) super.sizeInBytes() + (Partition_Response_Info_List_Size + partitionResponseInfoSize));
```

Comment:
I made a mistake, bufferToSend should never allocate `sizeInBytes()` bytes.
GetResponse is made of two parts, the metadata part and the data part.
`bufferToSend` should only has metadata part, since the data is carried by `toSend`.
if `toSend` doesn't support ByteBufHolder, then when writing GetResponse to network, we write `bufferToSend` first, and then `toSend` later. 
if `toSend` supports ByteBufHolder, then bufferToSend will be `CompositeByteBuf` later. This `CompositeByteBuf` contains metadata and data.